core-bundle/src/EventListener/PreviewUrlCreateListener.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\ArticleModel;
+use Contao\CoreBundle\Event\PreviewUrlCreateEvent;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\PageModel;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * @internal
+ */
+class PreviewUrlCreateListener
+{
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @var ContaoFramework
+     */
+    private $framework;
+
+    public function __construct(RequestStack $requestStack, ContaoFramework $framework)
+    {
+        $this->requestStack = $requestStack;
+        $this->framework = $framework;
+    }
+
+    /**
+     * Adds the page ID to the front end preview URL.
+     *
+     * @throws \RuntimeException
+     */
+    public function __invoke(PreviewUrlCreateEvent $event): void
+    {
+        if (!$this->framework->isInitialized() || !$event->getId() || !\in_array($event->getKey(), ['page', 'article'], true)) {
```

Comment:
Why the `!$event->getId()` check here? We are not checking this in the other listeners, either, and the event is always initialized with an ID.

Reply:
The idea is to exit early in order to prevent the unnecessary database call to `PageModel` or `ArticleModel` via `$adapter->findByPk` further down if there is no ID.

Reply:
You should IMHO inline the assignment then `… || !($id = $event->getId()) || …` and reuse it further down. :slightly_smiling_face:

Reply:
IMHO, we should remove the check completely, because the event is always initialized with an ID. Or at least add it to the other listeners as well.

Reply:
The `PreviewUrlCreateListener` in the news and calendar bundle try to get the ID from the request themselves. Thus, they do not necessarily rely on the ID of the event - on the contrary, they have to work with an empty ID (that's what the original issue was about).

https://github.com/contao/contao/blob/5b719bf9ee5191d4b87389398386b0da38b512bd/news-bundle/src/EventListener/PreviewUrlCreateListener.php#L67-L75

The generation of a preview url for a page or article however has no custom logic to retrieve the ID. With an empty ID or an ID of 0 the listener will not be able to generate an url and will just create `WHERE tl_page.id=0` requests to the database. That's why I assume an early exit is appropriate here.

Inlining the `$id` variable seems like a good idea.